Cypress

aguscha333 opened the pull request:
This PR integrates [Cypress](https://www.cypress.io/) into the base. For those who don't know what Cypress is, it's testing tool that allows you to easily implement not only unit tests but also integration and e2e.
 
The intention of this PR is to have a place where other developers can leave feedback on the implementation and documentation as well.

I still have some sections to complete in the readme but the most important ones have already a draft written.

Any questions that you have please let me know.

Here is a link to the [README](https://github.com/rootstrap/react-redux-base/tree/cypress/cypress) I added.

-----
### Update
Readme updated to be more complete and easy to understand.

To Do:
add support for e2e on the frontend side.

aguscha333 commented on cypress.json:
```diff
@@ -0,0 +1,12 @@
+  "baseUrl": "http://localhost:8000",
+  "indexUrl": "/",
+  "env": {
+    "API_URL": "https://test.com/api/v1"
```
I was able to configure cypress to use the env defined in process.env so it does not need to be repeated.

aguscha333 commented on src/utils/history.js:
```diff
@@ -8,6 +8,9 @@ class History {
     if (!instance) {
       instance = process.env.BROWSER ? createBrowserHistory() : createMemoryHistory();
     }
+    // if (window.Cypress) {
```
Nice catch! actually it was part of an attempt to solve something that didn't work so I'll remove it.

aguscha333 commented on cypress/integration/containers/LoginPage.spec.js:
```diff
@@ -0,0 +1,80 @@
+        .submit()
+        .wait('@loginStub');
+
+      cy.visit('/login');
```
Hmm I think that what it's actually being tested here is that after you logged in if you try going back to the login page you are always redirected to the home page.

Perhaps I should also assert that before that visit the url is also the homepage so we also test that after login user is redirected to the homepage.

At the same time it's kind of the same bit of code that redirects the user so we would be testing it twice.

```
  if (authenticated) {
    return <Redirect to={routes.index} />;
  }
```

WDYT?

felipegarcia92 replied:
I think I missed the test unit description, which already explains what you just explained in the comment. This was about the redirection, so it looks good. I don't think you need to extend this test or anything.

santiestra commented on cypress/README.md:
```diff
@@ -0,0 +1,300 @@
+ Parameter | Description | Values / Type
+ --------- | ----------- | -------------
+ name | this will be the stub identifier | **string**
+ method | Show file differences that haven't been staged | **HTTP METHODS** defined in constants
```
Is the description for `method` ok?

aguscha333 replied:
🙈 no hehe, copy & paste did me wrong, I'll fix it.

aguscha333 commented:
@santiestra regarding leaving Jest integration, I didn't remove jest completely. What I removed were the specs that we had since now those flows are tested in cypress. I also removed the scripts in package.json and replaced them with cypress ones.
I can add a script that is for example : `test:jest` that runs the old command so cypress is still the default.
I think that having the same test command run both is something we should take into consideration if we see that most projects end up using both. If not I think it's better to keep Jest as a separate command.

bernabe9 commented on cypress/README.md:
```diff
@@ -0,0 +1,300 @@
+----
+
+## Setup
+Cypress comes almost ready for you to use in this branch. There are a few configurations that you need to change in order to get the most out of it.
```
```suggestion
Cypress comes almost ready for you to use it. There are a few configurations that you need to change in order to get the most out of it.
```

Since we are going to merge this branch, I wouldn't mention it

aguscha333 replied:
Fixed, I switched it up a bit.

bernabe9 commented on package.json:
```diff
@@ -87,17 +92,22 @@
     "css-loader": "1.0.0",
+    "cypress": "3.3.0",
+    "cypress-image-snapshot": "3.0.2",
+    "cypress-react-unit-test": "2.4.1",
```
where are you using this package?

aguscha333 replied:
I am not. I added it when I was trying stuff and thought oh cool! you can write unit tests if you use this library.

I think it's definitely worth taking a look at it later but will remove it for now.
Perhaps we can write a unit test for the input component, we don't have many components in the base to showcase how it could be used.

bernabe9 commented on package.json:
```diff
@@ -87,17 +92,22 @@
     "enzyme-adapter-react-16": "1.5.0",
     "eslint": "4.19.1",
     "eslint-config-airbnb": "16.1.0",
+    "eslint-plugin-cypress": "2.2.1",
```
where are you using this package?

aguscha333 replied:
in eslintrc
```
  "plugins": [
    "react",
    "react-hooks",
    "cypress"
  ],
```

missx commented on cypress/README.md:
```diff
@@ -0,0 +1,300 @@
+-----
+
+## Running tests
+Running tests  is very easy, and we have already included some for you. To get a taste of how cypress works start your server with `ENV=cypress yarn start` and then run `yarn test:open`. The Test Runner will open and a list of tests to run will appear. Choose one of them and voilá, you will see screenshots of the test running in real time.
```
Two spaces after Running tests?

missx commented on cypress/README.md:
```diff
@@ -0,0 +1,300 @@
+
+For each stub you want to do, you will need to define certain parameters you want to pass to it.
+
+To keep stubs organized, there is a folder call `stubs` that should be the equivalent to `api` folder in `src`.
```
`called`

missx commented on cypress/README.md:
```diff
@@ -0,0 +1,300 @@
+ errors | Define errors to check. | **array of object**, objects inside follow **validation structure** | **false**
+ warnings | Define warnings to check. | **array of object**, objects inside follow **validation structure** | **false**
+
+Inside errors and warnings arrays you can define validations with this attributes:
```
Here attributes is plural, so `these` instead of `this`

aguscha333 replied:
🙈
